packages/flutter/lib/src/foundation/_isolates_io.dart

```diff
@@ -13,56 +13,81 @@ import 'isolates.dart' as isolates;
   debugLabel ??= kReleaseMode ? 'compute' : callback.toString();
+
   final Flow flow = Flow.begin();
   Timeline.startSync('$debugLabel: start', flow: flow);
-  final ReceivePort resultPort = ReceivePort();
-  final ReceivePort exitPort = ReceivePort();
-  final ReceivePort errorPort = ReceivePort();
-  Timeline.finishSync();
-  await Isolate.spawn<_IsolateConfiguration<Q, FutureOr<R>>>(
-    _spawn,
-    _IsolateConfiguration<Q, FutureOr<R>>(
-      callback,
-      message,
-      resultPort.sendPort,
-      debugLabel,
-      flow.id,
-    ),
-    errorsAreFatal: true,
-    onExit: exitPort.sendPort,
-    onError: errorPort.sendPort,
-  );
-  final Completer<R> result = Completer<R>();
-  errorPort.listen((dynamic errorData) {
-    assert(errorData is List<dynamic>);
-    if (errorData is List<dynamic>) {
-      assert(errorData.length == 2);
-      final Exception exception = Exception(errorData[0]);
-      final StackTrace stack = StackTrace.fromString(errorData[1] as String);
-      if (result.isCompleted) {
-        Zone.current.handleUncaughtError(exception, stack);
-      } else {
-        result.completeError(exception, stack);
-      }
-    }
-  });
-  exitPort.listen((dynamic exitData) {
-    if (!result.isCompleted) {
-      result.completeError(Exception('Isolate exited without result or error.'));
-    }
-  });
-  resultPort.listen((dynamic resultData) {
-    assert(resultData == null || resultData is R);
-    if (!result.isCompleted)
-      result.complete(resultData as R);
-  });
-  await result.future;
-  Timeline.startSync('$debugLabel: end', flow: Flow.end(flow.id));
-  resultPort.close();
-  exitPort.close();
-  errorPort.close();
+  final RawReceivePort port = RawReceivePort();
   Timeline.finishSync();
-  return result.future;
+
+  void _timeEndAndCleanup() {
+    Timeline.startSync('$debugLabel: end', flow: Flow.end(flow.id));
+    port.close();
+    Timeline.finishSync();
+  }
+
+  final Completer<dynamic> completer = Completer<dynamic>();
+  port.handler = (dynamic msg) {
+    _timeEndAndCleanup();
+    completer.complete(msg);
+  };
+
+  try {
+    await Isolate.spawn<_IsolateConfiguration<Q, R>>(
+      _spawn,
+      _IsolateConfiguration<Q, R>(
+        callback,
+        message,
+        port.sendPort,
+        debugLabel,
+        flow.id,
+      ),
+      errorsAreFatal: true,
+      onExit: port.sendPort,
+      onError: port.sendPort,
+      debugName: debugLabel,
+    );
+  } on Object {
+    _timeEndAndCleanup();
+    rethrow;
+  }
+
+  final dynamic response = await completer.future;
+  if(response == null) {
+    throw RemoteError('Isolate exited without result or error.', '');
+  }
+
+  assert(response is List<dynamic>);
+  response as List<dynamic>;
+
+  final int type = response.length;
+  assert(1 <= type && type <= 3);
+
+  switch (type) {
+    // success; see _buildSuccessResponse
+    case 1:
+      assert(response[0] is R);
+      return response[0] as R;
```

Comment:
is this redundant considering we do explicit `as` on the next line?

(same question to three asserts below)

Reply:
I added them because I wanted it to be clear it is our job to guarantee the type is `R`.

While I agree that they are redundant, I would assume if someone else read the code they wouldn't have to wonder and look at the surrounding code to make sure the `... as Type` casts are safe.

I also took for inspiration [the initial comment by @irhn in a dart sdk issue](https://github.com/dart-lang/sdk/issues/47164#issuecomment-926862699) where the assert was used to assert the response was a `List<dynamic>`. Which could also be argued in this pull request as redundant since an `as` cast follows it immediately.

I'm not personally attached and can remove these `assert`s but if possible I would prefer to keep them since *I think* they make the code clearer for a new reader (at least the intent).

Reply:
> FWIW, Flutter style encourages liberal use of asserts if they can be at all helpful for debugging:

What is mentioned in the doc makes sense, but `assert(v is T)` is built into `v as T`, produces basically exact same exception being thrown.

Not a big deal for sure though.

Reply:
Hypothetically:

```dart
void main(List<String> args) => args.isEmpty ? fn1(1) : fn2(1);

void fn1(dynamic args) {
  assert(args is List);
}

void fn2(dynamic args) => args as List;
```

running: `dart run --enable-asserts main.dart`  
produces: `... Failed assertion: line 4 pos 10: 'args is List': is not true. ...`

while `dart run --enable-asserts main.dart 1` 
produces: `... type 'int' is not a subtype of type 'List<dynamic>' in type cast ...`

Again, personally I prefer the assertion during development.
Because a type cast signals to me that I might not understand the program flow correctly or that I made the wrong assumption.
While an assertion to me means this was intended and that something somewhere changed which made this intended assumption incorrect.

*Also I always feel a bit anxious if I don't use an assertion before a type cast.* Even in cases like this where the code guaranteeing the assumption is close. But that might easily change in a future change.

An additional reason why I added the assertions is because I feel the most correct way to write that would have been:
```dart
final dynamic result = response[0];
if(result is R) {
  return result;
} else {
   throw Exception('Should never happen!!!');
}
```
But that code is an assertion! 
Which could be reduced to:
```dart
assert(response[0] is R);
return response[0];
```

But because the type of `response[0]` is inferred as `dynamic`. We need to add the explicit cast, and arrive at the code in the PR.

I feel, I'm making a big deal out of this rival detail. But just wanted to explain how I reasoned.